File: solutions/0/0/7/8/p0078_s0.c
/* Solution to Euler Problem 78: Coin Partitions. */
#include "runner.h"

/*
 * Bottom-up DP for the partition function p(n) via Euler's pentagonal number
 * theorem: p(n) = sum_{k != 0} (-1)^(k+1) p(n - g(k)), where g(k) = k(3k-1)/2
 * are the generalized pentagonal numbers. Only O(sqrt(n)) terms per n have a
 * non-negative argument, so building the table to the answer N is O(N*sqrt(N)).
 * Since we only need divisibility, every value is reduced mod divisor, keeping
 * all stored numbers bounded. The first n with p(n) == 0 (mod divisor) is the
 * answer.
 */
const char *solve(int argc, char *argv[]) {
    static char _answer[32];
    int divisor = parse_int(argv[1]);

    /* Allocate a growing array for partition values mod divisor */
    int capacity = 100000;
    long long *partitions = malloc((size_t)capacity * sizeof(long long));
    if (!partitions) {
        fprintf(stderr, "Out of memory\n");
        { snprintf(_answer, sizeof _answer, "%lld", (long long)(-1)); return _answer; }
    }
    partitions[0] = 1;  /* p(0) = 1 seeds the recurrence */

    int n = 1;
    while (1) {
        if (n >= capacity) {
            /* Doubling growth keeps reallocs at O(log N) for unknown answer size */
            capacity *= 2;
            long long *tmp = realloc(partitions, (size_t)capacity * sizeof(long long));
            if (!tmp) {
                fprintf(stderr, "Out of memory\n");
                free(partitions);
                { snprintf(_answer, sizeof _answer, "%lld", (long long)(-1)); return _answer; }
            }
            partitions = tmp;
        }

        long long pval = 0;
        int k = 1;
        while (1) {
            /* Process the (k, -k) pentagonal pair together; one sign covers both */
            long long pent_pos = (long long)k * (3 * k - 1) / 2;
            long long pent_neg = (long long)k * (3 * k + 1) / 2;

            /* pent_pos is the smaller member, so this is the loop's termination test */
            if (pent_pos > n) break;

            int sign = (k % 2 == 1) ? 1 : -1;

            pval += sign * partitions[n - pent_pos];

            /* pent_neg can exceed n while pent_pos does not, so guard it separately */
            if (pent_neg <= n) {
                pval += sign * partitions[n - pent_neg];
            }

            k++;
        }

        /* Subtractions may leave pval negative, so normalize into [0, divisor) */
        pval = ((pval % divisor) + divisor) % divisor;
        partitions[n] = pval;

        if (pval == 0) {
            free(partitions);
            { snprintf(_answer, sizeof _answer, "%lld", (long long)((long long)n)); return _answer; }
        }

        n++;
    }
}

File: solutions/0/0/7/8/p0078_s0.py
#!/usr/bin/env python3.14
# -*- coding: utf-8 -*-
""" Solution to Euler Problem 78: Coin Partitions [Level 3]. """
from __future__ import annotations

import functools
import itertools

from solver.runners import runner


@functools.lru_cache(maxsize=None)
def pentagonal(x: int) -> int:
    """Return the generalized pentagonal number g(x) = x(3x-1)/2."""
    return x * (3 * x - 1) // 2


def least_number_with_partitions_divisible_by(divisor: int) -> int:
    """Bottom-up DP for p(n) via Euler's pentagonal recurrence, reduced mod divisor; O(N*sqrt(N)).

    Each n sums (-1)^(k-1) p(n - g(k)) over the (k, -k) pentagonal pairs, stopping once g(k) > n
    (only O(sqrt(n)) terms contribute). The first n with p(n) == 0 (mod divisor) is returned.
    """
    partitions = [1]
    for n in itertools.count(1):
        partition_value = 0
        k = 1
        while True:
            pent_k1 = pentagonal(k)
            pent_k2 = pentagonal(-k)
            if pent_k1 > n:
                break
            partition_value += (-1) ** (k - 1) * partitions[n - pent_k1]
            if 0 < pent_k2 <= n:
                partition_value += (-1) ** (k - 1) * partitions[n - pent_k2]
            k += 1
        partition_value %= divisor
        partitions.append(partition_value)
        if partition_value == 0:
            return n
    return -1


@runner.main
def solve(*args: str) -> str:
    """Parse the divisor and return the least n with p(n) divisible by it."""
    divisor = runner.parse_int(args[0])

    return str(least_number_with_partitions_divisible_by(divisor=divisor))


if __name__ == "__main__":
    raise SystemExit(solve())
